### DISCUSSION TITLE
"Inverse literate" configuration

### DISCUSSION THREAD

**Discussion by shombando** [nobiot/org-transclusion#114] [Show and tell]
@nobiot thanks for creating this great package. I wanted to share an example of how I'm using it. I'm using it to create an "inverse literate" configuration for Emacs. Essentially, instead of creating code blocks in an org file and tangling them out, the org file has org-transclusion code blocks to pull in the specific parts of the config. 

I [wrote a bit more](https://shom.dev/posts/20211230_inverse-literate-config-via-org-transclusion/) about my motivation behind it with a short example/explanation if anyone is interested.  Or you can [jump straight to my config](https://git.sr.ht/~shom/.emacs.custom) and take a look. 

I haven't finished writing about all the packages I use yet, but the good thing is that the org file doesn't need to talk every line/block you have in your config only what you want to highlight, which is another "advantage" than a literate config that is tangled out. 

Thanks again for the great package, thought I'd share a perhaps unintended use case that is working very well for me!

**Comment by nobiot** [comment]
@shombando

Thank you for sharing how you use Org-transclusion!
When it comes to literate configuration, I think I'm in the same camp as you. I tried with tangling but I decided that I preferred changing Elisp code directly rather than via source blocks in an Org file and to go back to `.el` files with comments. Your "inverse literate" makes sense to me.

> a perhaps unintended use case that is working very well for me!

I have always thought that if a tool is good, there will be people who put it in good use in ways that its creator has never thought of. So your use case means a lot to me :) Thank you.

On a side note, I really like what you are doing with gemini protocol with your capsule/blog and EmacsVanillaChocolateSwirl. The Emacs wizard idea by Yuan looks also good. I wonder if we/you/I/someone can pick it up from where it was left...

**Comment by shombando** [comment]
Thanks for the gracious sentiment @nobiot, very kind of you!  I went a step further and added an in-buffer function (if that's the right term) to automatically generate the readme.md each time the org file is saved: https://git.sr.ht/~shom/.emacs.custom#exporting-readme-markdown (please let me know if you think there's a better way to do this).

I definitely want to make some time and come back to automating gemini publishing from my single org file that publishes using ox-hugo. As for EmacsVanillaChocolateSwirl it's a (tasty) mouthful of a project name. I have learned a lot more since I started that but I'm really glad I have that "new user perspective" documented. I think I want to convert all of that over to inverse literate, so new users can just hack on the .el file and remove another layer of abstraction while still providing the commentary. What do you think?

I would love for Emacs to have a new user wizard (which only triggers if a config isn't detected) but my emacs-lisp chops are non-existent. But I would love to contribute to an effort like that (I can hack around if there's a structure). If you ever consider it, please let me know.

**Comment by devcarbon-com** [comment]
I'm planning to use this same approach, and am hoping to make a very streamlined workflow with as little friction as possible. I'll update here when I feel like I have something that is starting to be coherent. :)

**Reply by nobiot** [reply]
Cool 😎

**Reply by shombando** [reply]
@devcarbon-com cool, looking forward to seeing what you do and learning from it!

**Comment by devcarbon-com** [comment]
@nobiot I'm curious, is there a technical limitation that prevents us from using live sync on source code blocks?  If it's possible, do you have a rough idea how much effort it will take?

**Reply by nobiot** [reply]
Yes, the technical issue exists for the current design of org-transclusion. It uses `org-element.el`, which is the built-in parser of `org`, which lets us do the on-the-fly filtering to represent the source org buffer differently in the transclusion target. 

`org-transclusoin-live-sync` relies on the two buffers (source and target) to be **strictly** identical; not a single space can be different. In addition to this, live-sync needs to know the location (start and end points) of the source. To get the content, live-sync disables the filters and get the source element as it is in the source buffer. For the original location, `org-element.el` does not provide the information of the "origin". I have tried various ways to add the origin in the property; the added custom property does not seem to be carried over in the process. `org-element` in the end uses `buffer-string-no-properties` (or `substring-no-properties`), which removes it. 

Now... I extensively looked at this situation two years ago for Org 9.3 and 9.4 at the time I developed `org-transclusion`. Things might have changed (I haven't looked at the changes in this space in 9.6 yet). I haven't brought this as a request to the org-development mailing list -- they may be receptive to this idea.

I think I have also tried overriding the `*-no-properties` function to test the idea -- I don't seem to have recorded it properly (I'm looking at my dev log), but I recall that it worked. But you really would not want to meddle with such a generic and common function -- You could use `cl-letf` or `cl-flet` to localize this modification to the current scope. Perhaps that's one way to go, but I'd first discuss this matter in the mailing list (I don't have the bandwidth at the moment, so it will be great if you could do this...).

For the elements for which live-sync works, somehow they all have a parent element, which has the information of the original location. The elements that live-sync does not work do not have such parent element. I manually tested these and recorded in the document string of `org-transclusion-live-sync-start`:

```
For transclusions of Org elements or a buffer, live-sync works
only on the following elements: center-block, drawer,
dynamic-block, latex-environment, paragraph, plain-list,
quote-block, special-block table, and verse-block.

It is known that live-sync does not work for the other Org
elements: comment-block, export-block, example-block,
fixed-width, keyword, src-block, and property-drawerd.
```

**Reply by ParetoOptimalDev** [reply]
I asked for some help in the org matrix room since I was already in it.

> have tried various ways to add the origin in the property; the added custom property does not seem to be carried over in the process.

Might be answered by:

> Only keywords listed in org-element-affiliated-keywords and keywords starting from ATTR_ are considered affilieated with the following element 

and a potential solution:

> you can try to use #+ATTR_TRANSCLUDE: :origin <foo> 

Here's the full log in case it's useful.

<details>
<summary> full log</summary>

```
  pareto-opti…​Does anyone know of a way to track the "origin" of a transcluded org source block? Adding custom properties didn't seem to work for the author. context: https://github.com/nobiot/org-transclusion/discussions/114#discussioncomment-5201326 
  pareto-opti…​in org 9.4 because org-element uses `buffer-string-no-properties`, so perhaps a change from then to 9.6 now might allow you to preserve properties in parsed org elements?  
  yantar92 (Org contributor)​In reply to @pareto-optimal-dev:matrix.org
Does anyone know of a way to track the "origin" of a transcluded org source block? Adding custom properties didn't seem to work for the author. context: https://github.com/nobiot/org-transclusion/discussions/114#discussioncomment-5201326

What do you mean by origin? 
 12:06 
  p00f​In reply to @chda:matrix.org
org-html-export-to-html doesn't show images which i inserted using org-download-clipboard

i had to manually change the img src to add a ./ before the path to get firefox to recognize it, can i change this in org settings 
 13:32 
  pareto-optimal-dev​In reply to yantar92 (Org contributor) 

In reply to @pareto-optimal-dev:matrix.org
Does anyone know of a way to track the "origin" of a transcluded org source block? Adding custom properties didn't seem to work for the author. context: https://github.com/nobiot/org-transclusion/discussions/114#discussioncomment-5201326

What do you mean by origin? 

A custom property named origin. To differentiate between the orginal source block and transcluded ones. Here's an example org file and how it's rendered using org-exclusion that might explain: https://github.com/nobiot/org-transclusion/issues/140#issuecomment-1456505804 
  yantar92 (Org contributor)​In reply to @pareto-optimal-dev:matrix.org
A custom property named origin. To differentiate between the orginal source block and transcluded ones. Here's an example org file and how it's rendered using org-exclusion that might explain: https://github.com/nobiot/org-transclusion/issues/140#issuecomment-1456505804

Only keywords listed in org-element-affiliated-keywords and keywords starting from ATTR_ are considered affilieated with the following element 
  yantar92 (Org contributor)​you can try to use #+ATTR_TRANSCLUDE: :origin <foo> 
```
</details>

**Reply by nobiot** [reply]
Thank you for asking the question in the matrix. I am not sure how this can be applied to our case, though. How do we get the start and end point of an element with this way?

**Reply by ParetoOptimalDev** [reply]
I'm low on context honestly and will likely need to play around with examples in emacs for more, but after reading:

> For the original location, org-element.el does not provide the information of the "origin".
> I have tried various ways to add the origin in the property; the added custom property does not seem to be carried over in the process.

I assumed that you could fix the problem of the custom added property not being carried over by making sure you prepend it with `ATTR_`. 

So to answer your question, add the origin in whatever what you tried to add it before but make sure the property name is prepended with `ATTR_`.

**Reply by ParetoOptimalDev** [reply]
For totally unrelated reasons I was looking into polymode which somewhat surprisingly might be of help to this problem.

From my understanding it has to differentiate between different types of code within a document. So it replaces org source blocks as I undersand it with:

> Spans
> span is a contiguous and homogeneous region of text of the same Emacs major mode. There are 4 types of spans: head, body, tail and host.
All spans are closed on the left and open on the right. For example, a span spanning [3, 5) contains character at position 3 but not at position 5. This is consistent with how emacs' functions deal with text properties.

https://polymode.github.io/concepts/

So the polymode implementation has the same requirements for identifying positioning that one would need to transclude org source blocks it seems.

@vspinu If you have the time and interest, can you verify or comment on the above?

Probably getting ahead of myself, but if this is the case and the notion of spans, chunks, etc that polymode have work for both use cases perhaps they belong in emacs core?

**Reply by devcarbon-com** [reply]
@nobiot Sure.  Can you point me to the relevant functions that use `*-no-properties` so that I can be specific in my conversation with them?

**Reply by devcarbon-com** [reply]
Ah I think I found it. `org-element-src-block-parser` and friends.

**Reply by devcarbon-com** [reply]
Hhm, I think org might have changed such that it works now. 

I have a monkey-patched local version of org-transclusion working right now with live-sync on src-blocks. 

I'm using org version 9.7-pre, so I don't know for how long the extra properties have been preserved.

The monkey patching was because `org-transclusion-beg-mkr` and `org-transclusion-end-mkr` are inclusive of the `#begin_src` and `#end_src` markers, so the content length was offset.

**Reply by devcarbon-com** [reply]
Experimental branch here: https://github.com/devcarbon-com/org-transclusion/tree/feature--org-block-live-sync

**Comment by reyman** [comment]
Really interesting thread, i'm very interested by such workflow. 

In this issue ( https://github.com/nobiot/org-transclusion/issues/144), i also try to think about the "inverse literate" configuration in a "sharing code source in a team workflow".  

When many people use the same repository, with or without org-mode, they change name of function, line number, etc. 

Detecting the code source to transclude using generic "uuid" (like org-roam finally) is probably more robust in a team environnement. 

```
<< uuid xcfvdes34d 
def myFunctionA() : 
    print ("helloworld")
    ... 
>>
```

**Reply by devcarbon-com** [reply]
@reyman I'm getting closer to releasing a prototype.  It's been fun, but I'm also a bit limited in the time I have available to invest in this. 

So far I've got (among other things)
- Automatic uuid insertion
- an easy-to-use transient UI
- seamless jumping from code to .org and vise-versa

There's still plenty more on the todo list, and currently it's best tuned for working on lisps.

**Reply by ParetoOptimalDev** [reply]
Nice job!

I fear if your goal is collaborating with people that update the non-literate code though... I've found any sort of annotation even if minimal makes literate<-->non-literate codebase interaction a non-starter or totally prevents adoption. Or at least makes one very unpopular :sweat_smile: 

However... if you are interested there is a solution that's been used by the leo editor for some time to enable sync between literate/non-literate without any annotations or "sentinels":

> The algorithm works with any kind of text file. The algorithm uses only difflib. It knows nothing about the text or its meaning. No parsing is ever done.
>
> Suppose file x has been changed outside of Leo. When Leo reads x it does the following:
>
> Recreates the old version of x without sentinels by writing the @clean x outline into a string, as if it were writing the @clean x outline again.
> 
> Recreates all the lines of x with sentinels by writing the @clean x outline into a string, as if it was writing an @file node! Let’s call these lines the old sentinels lines.
> 
> Uses difflib.SequenceMatcher to create a set of diffs between the old and new versions of x without sentinels.
>
> Terminology: the diffs tell how to change file a into file b. The actual code uses this terminology: a is set of lines in the old version of x, b is the set of lines in the new version of x.
> 
> Creates a set of lines, the new sentinels lines using the old sentinels lines, the a and b lines and the diffs.
> 
> This is the magic. Bernhard Mulder’s genius was conceiving that a three-way merge of lines could produce the new outline, with sentinels. The code is in x.propagate_changed_lines and its helpers.
> 
> Replaces the @clean tree with the new tree created by reading the new sentinels lines with the @file read logic.
> 
> Important: The update algorithm never changes sentinels. It never inserts or deletes nodes. The user is responsible for creating nodes to hold new lines, or for deleting nodes that become empty as the result of deleting lines.

http://leoeditor.com/appendices.html#the-mulder-ream-update-algorithm

**Reply by devcarbon-com** [reply]
@ParetoOptimalDev, interesting! 

I'm not sure how I'd go about implementing such an algorithm for emacs, but I'd be open to suggestions/pull-requests.  

The way I'm going about it for my codebase is to try to make the id as unobtrusive as possible. 

It looks like this:

```elisp
(defun marker-at-line (line) ;<id:1672242453>
  "Create a marker at the beginning of LINE in the current buffer."
  (set-marker (make-marker) (line-beginning-position line)))
```

And it's set to transclude the sexp at point. (or rather, at the beginning of the line excluding white-space.)

**Reply by nobiot** [reply]
@devcarbon-com I'll be honest and I don't think I can follow the discussion (it's fascinating but the content is beyond me). Is this related to your PR #157? Should we somehow collectively review it?

If this is OK, I'd love others like @ParetoOptimalDev and @reyman to review @devcarbon-com's PR #157.

I am no developer, and I don't think I am very useful for this PR -- the feature is pretty much geared toward developer audience, I think.

Let me know what you think.

Also... As an idea, perhaps @devcarbon-com or someone with interest in the developer usage might like to own this part of the development? (I would love someone like @devcarbon-com with enthusiasm and skills to take the lead). The key functions should be in a single file `org-transclusion-src-lines.el` (with occasional change to the core `org-transclusion.el`, whose dicsussion I can also participate in) so the separation of ownership and focus should be feasible, in my opinion. Not sure how this might sound to you -- far-fetched? fair?

**Reply by ParetoOptimalDev** [reply]
When I get some time I will review #157 further, but a quick glance both the idea and the implementation seem good. I'll comment more in there.

Having someone interested in developer usage owning that part of it sounds like a great idea. I wouldn't be able to because I've not yet been able to integrate org-transclusion into important code in my work or personal life, but that's becoming possible/desireable :smile:

**Comment by joshcho** [comment]
I have taken a keen interest in using transclusion for development workflow. I do full-stack development with Clojure, and I often need to jump between different sets of files. Suppose on the frontend, I have files A, B, C, and on the backend, D, E, F. It's hard to juggle different windows for different kinds of workflows. But with org-transclusion, I can have transcluded A, B, C in a frontend.org file, and simulate the experience of having three vertically stacked windows in a single window. The advantage of transclusion is that "toggling" the visibility of individual files is much easier, not to mention all the org-mode advantages you get.

I think with some :lines trickery you can also change the "view" into the files themselves in order to simulate cursor move in a buffer. A src code live-sync on top of this would be great.
